File: tools/image_processing/trim_tiles.js
#!/usr/bin/env node

/**
 * Node.js版本：自动裁剪PNG图片的透明边缘
 * 备选方案（如果Python不可用）
 *
 * 安装依赖: npm install sharp
 * 运行: node trim_tiles.js
 */

const fs = require('fs');
const path = require('path');
const { promisify } = require('util');

// 检查sharp库
let sharp;
try {
    sharp = require('sharp');
} catch (e) {
    console.error('❌ 错误: sharp库未安装');
    console.error('请先安装: npm install sharp');
    process.exit(1);
}

const readdir = promisify(fs.readdir);
const stat = promisify(fs.stat);
const copyFile = promisify(fs.copyFile);
const mkdir = promisify(fs.mkdir);

class TileImageProcessor {
    constructor(tilesDir) {
        this.tilesDir = tilesDir;
        this.backupDir = path.join(
            path.dirname(tilesDir),
            'backup',
            new Date().toISOString().slice(0, 19).replace(/:/g, '')
        );
        this.report = {
            timestamp: new Date().toISOString(),
            tiles_dir: tilesDir,
            total_processed: 0,
            total_failed: 0,
            tiles: []
        };
    }

    async backupOriginal(pngPath) {
        try {
            await mkdir(this.backupDir, { recursive: true });
            const backupPath = path.join(this.backupDir, path.basename(pngPath));
            await copyFile(pngPath, backupPath);
            return backupPath;
        } catch (e) {
            console.error(`❌ 备份失败: ${path.basename(pngPath)} - ${e.message}`);
            return null;
        }
    }

    async trimTransparentEdges(pngPath) {
        try {
            const image = sharp(pngPath);
            const metadata = await image.metadata();
            const originalSize = {
                width: metadata.width,
                height: metadata.height
            };

            // 获取原始文件大小
            const originalBytes = (await promisify(fs.stat)(pngPath)).size;

            // 裁剪透明边缘
            const { data: trimmedData, info: trimmedInfo } = await image
                .trim()
                .toBuffer({ resolveWithObject: true });

            const trimmedSize = {
                width: trimmedInfo.width,
                height: trimmedInfo.height
            };

            // 保存裁剪后的图片
            await fs.promises.writeFile(pngPath, trimmedData);
            const trimmedBytes = trimmedData.length;

            // 计算裁剪的像素
            const croppedPixels = {
                left: 0,
                top: 0,
                right: 0,
                bottom: 0
            };

            // 计算透明比例
            const totalPixels = originalSize.width * originalSize.height;
            const contentPixels = trimmedSize.width * trimmedSize.height;
            const transparencyRatio = 1.0 - (contentPixels / totalPixels);

            return {
                success: true,
                original_size: originalSize,
                trimmed_size: trimmedSize,
                cropped_pixels: croppedPixels,
                transparency_ratio: `${(transparencyRatio * 100).toFixed(1)}%`,
                file_size: [originalBytes / 1024, trimmedBytes / 1024],
                error: null
            };
        } catch (e) {
            return {
                success: false,
                original_size: null,
                trimmed_size: null,
                cropped_pixels: null,
                transparency_ratio: null,
                file_size: null,
                error: e.message
            };
        }
    }

    async processAllTiles() {
        if (!fs.existsSync(this.tilesDir)) {
            console.error(`❌ 目录不存在: ${this.tilesDir}`);
            return false;
        }

        const files = await readdir(this.tilesDir);
        const pngFiles = files
            .filter(f => f.startsWith('tile_') && f.endsWith('.png'))
            .sort();

        if (pngFiles.length === 0) {
            console.warn(`⚠️  未找到PNG文件在: ${this.tilesDir}`);
            return false;
        }

        console.log(`\n🔍 找到 ${pngFiles.length} 个PNG文件，开始处理...\n`);

        let maxWidth = 0;
        let maxHeight = 0;

        for (const file of pngFiles) {
            const pngPath = path.join(this.tilesDir, file);
            console.log(`处理: ${file}`);

            // 备份原始文件
            const backupPath = await this.backupOriginal(pngPath);
            if (backupPath) {
                console.log(`  ✓ 已备份到: ${backupPath}`);
            }

            // 裁剪透明边缘
            const result = await this.trimTransparentEdges(pngPath);

            if (result.success) {
                console.log(`  ✓ 裁剪成功`);
                console.log(`    原始尺寸: ${result.original_size.width}×${result.original_size.height}px`);
                console.log(`    裁剪后:   ${result.trimmed_size.width}×${result.trimmed_size.height}px`);
                console.log(`    透明比例: ${result.transparency_ratio}`);
                console.log(`    文件大小: ${result.file_size[0].toFixed(1)}KB → ${result.file_size[1].toFixed(1)}KB`);

                maxWidth = Math.max(maxWidth, result.trimmed_size.width);
                maxHeight = Math.max(maxHeight, result.trimmed_size.height);

                this.report.total_processed++;
            } else {
                console.log(`  ❌ 裁剪失败: ${result.error}`);
                this.report.total_failed++;
            }

            this.report.tiles.push({
                file: file,
                result: result
            });
            console.log();
        }

        this.report.max_trimmed_size = {
            width: maxWidth,
            height: maxHeight,
            note: '所有卡片裁剪后的最大尺寸'
        };

        this.report.summary = {
            total: pngFiles.length,
            success: this.report.total_processed,
            failed: this.report.total_failed
        };

        return this.report.total_failed === 0;
    }

    async saveReport(reportPath = null) {
        if (!reportPath) {
            reportPath = path.join(path.dirname(this.tilesDir), 'trim_report.json');
        }

        await fs.promises.writeFile(reportPath, JSON.stringify(this.report, null, 2), 'utf-8');
        console.log(`📋 处理报告已保存: ${reportPath}`);
        return reportPath;
    }

    printSummary() {
        console.log('\n' + '='.repeat(60));
        console.log('📊 处理摘要');
        console.log('='.repeat(60));
        console.log(`总数:     ${this.report.summary.total} 个文件`);
        console.log(`成功:     ${this.report.summary.success} ✓`);
        console.log(`失败:     ${this.report.summary.failed} ✗`);

        if (this.report.total_processed > 0) {
            console.log(`\n最大裁剪后尺寸: ${this.report.max_trimmed_size.width}×${this.report.max_trimmed_size.height}px`);
            console.log(`备份目录:       ${this.backupDir}`);
        }

        console.log('='.repeat(60) + '\n');
    }
}

async function main() {
    const tilesDir = path.join(
        __dirname,
        '../../src/cocos/assets/bundle/tiles'
    );

    console.log(`
╔════════════════════════════════════════════════════════════╗
║        Cocos Creator 卡片PNG透明边缘裁剪工具 (Node.js)     ║
║                  Tile Image Trimmer v1.0                   ║
╚════════════════════════════════════════════════════════════╝
    `);

    const processor = new TileImageProcessor(tilesDir);

    try {
        const success = await processor.processAllTiles();
        await processor.saveReport();
        processor.printSummary();

        process.exit(success ? 0 : 1);
    } catch (e) {
        console.error('❌ 处理失败:', e.message);
        process.exit(1);
    }
}

main();

File: tools/image_processing/trim_tiles.py
#!/usr/bin/env python3
"""
自动裁剪PNG图片的透明边缘
用于去除Cocos Creator卡片图片的透明填充区域
"""

import os
import sys
from pathlib import Path
from PIL import Image
import json
from datetime import datetime


class TileImageProcessor:
    def __init__(self, tiles_dir):
        """初始化处理器"""
        self.tiles_dir = Path(tiles_dir)
        self.backup_dir = self.tiles_dir.parent / "backup" / datetime.now().strftime("%Y%m%d_%H%M%S")
        self.report = {
            "timestamp": datetime.now().isoformat(),
            "tiles_dir": str(self.tiles_dir),
            "total_processed": 0,
            "total_failed": 0,
            "tiles": []
        }

    def backup_original(self, png_path):
        """备份原始PNG文件"""
        try:
            self.backup_dir.mkdir(parents=True, exist_ok=True)
            backup_path = self.backup_dir / png_path.name

            # 读取原始文件
            with open(png_path, 'rb') as src:
                with open(backup_path, 'wb') as dst:
                    dst.write(src.read())

            return str(backup_path)
        except Exception as e:
            print(f"❌ 备份失败: {png_path.name} - {e}")
            return None

    def trim_transparent_edges(self, png_path):
        """
        裁剪PNG图片的透明边缘

        返回: {
            'success': bool,
            'original_size': (width, height),
            'trimmed_size': (width, height),
            'cropped_pixels': {top, bottom, left, right},
            'transparency_ratio': float (0-1),
            'file_size': (original_kb, trimmed_kb),
            'error': str or None
        }
        """
        try:
            # 打开图片
            img = Image.open(png_path).convert("RGBA")
            original_size = img.size
            original_bytes = os.path.getsize(png_path)

            # 获取非透明区域的边界框
            bbox = img.getbbox()

            if bbox is None:
                # 全透明图片
                return {
                    'success': False,
                    'original_size': original_size,
                    'trimmed_size': None,
                    'cropped_pixels': None,
                    'transparency_ratio': 1.0,
                    'file_size': (original_bytes / 1024, 0),
                    'error': '图片全透明，无法处理'
                }

            # 计算裁剪信息
            left, top, right, bottom = bbox
            trimmed_size = (right - left, bottom - top)
            cropped_pixels = {
                'left': left,
                'top': top,
                'right': original_size[0] - right,
                'bottom': original_size[1] - bottom
            }

            # 计算透明比例
            total_pixels = original_size[0] * original_size[1]
            content_pixels = trimmed_size[0] * trimmed_size[1]
            transparency_ratio = 1.0 - (content_pixels / total_pixels)

            # 裁剪图片
            cropped_img = img.crop(bbox)

            # 保存回原位置
            cropped_img.save(png_path, "PNG", optimize=True)
            trimmed_bytes = os.path.getsize(png_path)

            return {
                'success': True,
                'original_size': original_size,
                'trimmed_size': trimmed_size,
                'cropped_pixels': cropped_pixels,
                'transparency_ratio': f"{transparency_ratio * 100:.1f}%",
                'file_size': (original_bytes / 1024, trimmed_bytes / 1024),
                'error': None
            }

        except Exception as e:
            return {
                'success': False,
                'original_size': None,
                'trimmed_size': None,
                'cropped_pixels': None,
                'transparency_ratio': None,
                'file_size': None,
                'error': str(e)
            }

    def process_all_tiles(self):
        """处理所有PNG文件"""
        if not self.tiles_dir.exists():
            print(f"❌ 目录不存在: {self.tiles_dir}")
            return False

        # 找到所有PNG文件
        png_files = list(self.tiles_dir.glob("tile_*.png"))

        if not png_files:
            print(f"⚠️  未找到PNG文件在: {self.tiles_dir}")
            return False

        print(f"\n🔍 找到 {len(png_files)} 个PNG文件，开始处理...\n")

        max_width = 0
        max_height = 0

        for png_path in sorted(png_files):
            print(f"处理: {png_path.name}")

            # 备份原始文件
            backup_path = self.backup_original(png_path)
            if backup_path:
                print(f"  ✓ 已备份到: {backup_path}")

            # 裁剪透明边缘
            result = self.trim_transparent_edges(png_path)

            if result['success']:
                print(f"  ✓ 裁剪成功")
                print(f"    原始尺寸: {result['original_size'][0]}×{result['original_size'][1]}px")
                print(f"    裁剪后:   {result['trimmed_size'][0]}×{result['trimmed_size'][1]}px")
                print(f"    透明比例: {result['transparency_ratio']}")
                print(f"    文件大小: {result['file_size'][0]:.1f}KB → {result['file_size'][1]:.1f}KB")

                # 记录最大尺寸（用于统一调整）
                max_width = max(max_width, result['trimmed_size'][0])
                max_height = max(max_height, result['trimmed_size'][1])

                self.report['total_processed'] += 1
            else:
                print(f"  ❌ 裁剪失败: {result['error']}")
                self.report['total_failed'] += 1

            self.report['tiles'].append({
                'file': png_path.name,
                'result': result
            })
            print()

        # 生成处理报告
        self.report['max_trimmed_size'] = {
            'width': max_width,
            'height': max_height,
            'note': '所有卡片裁剪后的最大尺寸'
        }
        self.report['summary'] = {
            'total': len(png_files),
            'success': self.report['total_processed'],
            'failed': self.report['total_failed']
        }

        return self.report['total_failed'] == 0

    def save_report(self, report_path=None):
        """保存处理报告"""
        if report_path is None:
            report_path = self.tiles_dir.parent / "trim_report.json"

        with open(report_path, 'w', encoding='utf-8') as f:
            json.dump(self.report, f, ensure_ascii=False, indent=2)

        print(f"📋 处理报告已保存: {report_path}")
        return report_path

    def print_summary(self):
        """打印处理摘要"""
        print("\n" + "="*60)
        print("📊 处理摘要")
        print("="*60)
        print(f"总数:     {self.report['summary']['total']} 个文件")
        print(f"成功:     {self.report['summary']['success']} ✓")
        print(f"失败:     {self.report['summary']['failed']} ✗")

        if self.report['total_processed'] > 0:
            print(f"\n最大裁剪后尺寸: {self.report['max_trimmed_size']['width']}×{self.report['max_trimmed_size']['height']}px")
            print(f"备份目录:       {self.backup_dir}")

        print("="*60 + "\n")


def main():
    """主函数"""
    # 默认处理路径
    tiles_dir = Path(__file__).parent.parent.parent / "src/cocos/assets/bundle/tiles"

    # 如果命令行指定了路径，使用命令行参数
    if len(sys.argv) > 1:
        tiles_dir = Path(sys.argv[1])

    print(f"""
╔════════════════════════════════════════════════════════════╗
║           Cocos Creator 卡片PNG透明边缘裁剪工具            ║
║                  Tile Image Trimmer v1.0                   ║
╚════════════════════════════════════════════════════════════╝
    """)

    processor = TileImageProcessor(tiles_dir)

    # 处理所有PNG
    success = processor.process_all_tiles()

    # 保存报告
    processor.save_report()

    # 打印摘要
    processor.print_summary()

    return 0 if success else 1


if __name__ == "__main__":
    sys.exit(main())
